File: pac-man.py
import random

N, M = 6, 15
counter = 0
a, b = -1, -1
grid = [['+' for x in range(M)] for y in range(N)]


# This function prints the grid of Gomoku as the game progresses
def print_grid():
    print('--' + '---' * M + '--')
    for i in range(N):
        print(end='|  ')
        for j in range(M):
            print(grid[i][j], end='  ')
        print(end='|')
        print()
    print('--' + '---' * M + '--')


# This function checks if the game has a win state or not
def check_win():
    if counter == N * M - (N * M * 1 // 6 + N * M * 1 // 8 + 1):
        return True
    else:
        return False


def check_available_move(i):
    if i == 3:
        if b + 1 > M - 1:
            return False  # right
    elif i == 1:
        if a - 1 < 0:
            return False  # up
    elif i == 2:
        if a + 1 > N - 1:
            return False  # down
    elif i == 4:
        if b - 1 < 0:
            return False  # left
    return True


# This function checks if the game has a tie state or not for the given mark
def check_lose(i):
    if i == 3:
        if grid[a][b + 1] == 'G':
            return True
    elif i == 1:
        if grid[a - 1][b] == 'G':
            return True
    elif i == 2:
        if grid[a + 1][b] == 'G':
            return True
    elif i == 4:
        if grid[a][b - 1] == 'G':
            return True
    return False


def check_wall(i):
    if i == 3:
        if grid[a][b + 1] == '|':
            return True
    elif i == 1:
        if grid[a - 1][b] == '|':
            return True
    elif i == 2:
        if grid[a + 1][b] == '|':
            return True
    elif i == 4:
        if grid[a][b - 1] == '|':
            return True
    return False


def move(i):
    global a
    global b
    global counter
    grid[a][b] = '.'
    if i == 3:  # right
        if grid[a][b + 1] == '+':
            counter += 1
            grid[a][b + 1] = 'P'
        elif grid[a][b + 1] == '.':
            grid[a][b + 1] = 'P'
        b += 1

    elif i == 1:  # up
        if grid[a - 1][b] == '+':
            counter += 1
            grid[a - 1][b] = 'P'
        elif grid[a - 1][b] == '.':
            grid[a - 1][b] = 'P'
        a -= 1

    elif i == 2:  # down
        if grid[a + 1][b] == '+':
            counter += 1
            grid[a + 1][b] = 'P'
        elif grid[a + 1][b] == '.':
            grid[a + 1][b] = 'P'
        a += 1

    elif i == 4:  # left
        if grid[a][b - 1] == '+':
            counter += 1
            grid[a][b - 1] = 'P'
        elif grid[a][b - 1] == '.':
            grid[a][b - 1] = 'P'
        b -= 1


# This function generates pac man
def generate_pac_man():
    global a
    global b
    a = random.randint(0, N - 1)
    b = random.randint(0, M - 1)
    grid[a][b] = 'P'


# This function generates ghosts
def generate_ghost():
    f = N * M * 1 // 6
    while f > 0:
        a = random.randint(0, N - 1)
        b = random.randint(0, M - 1)
        while grid[a][b] == 'P':
            a = random.randint(0, N - 1)
            b = random.randint(0, M - 1)
        grid[a][b] = 'G'
        f -= 1

    # This function generates a wall


def generate_wall():
    f = N * M * 1 // 8
    while f > 0:
        a = random.randint(0, N - 1)
        b = random.randint(0, M - 1)
        while grid[a][b] == 'P' or grid[a][b] == 'G':
            a = random.randint(0, N - 1)
            b = random.randint(0, M - 1)
        grid[a][b] = '|'
        f -= 1


# This function checks if given position is valid or not
def check_valid_direction(i):
    if i == 1 or i == 2 or i == 3 or i == 4:
        return True
    else:
        return False


# This function clears the game structures
def grid_clear():
    global grid
    grid = [['+' for x in range(M)] for y in range(N)]
    global counter
    counter = 0


# This function reads a valid position input
def read_input():
    i = int(input('Enter the direction: '))
    while not check_available_move(i) or not check_valid_direction(i):
        i = int(input('Enter a valid direction: '))
    return i


# MAIN FUNCTION
def play_game():
    print("Pac_Man Game!")
    print("Welcome...")
    print("============================")
    while True:
        # Prints the grid
        print_grid()
        i = read_input()
        # Check if the grid has a tie state
        if check_lose(i):
            # Prints the grid
            print_grid()
            # Announcement of the final statement
            print("GAME OVER!")
            break
        if check_wall(i):
            print("WHATCH OUT you hit a wall !! please ", end=' ')
            i = read_input()
        # Move with the input direction
        move(i)
        # Check if the state of the grid has a win state
        if check_win():
            # Prints the grid
            print_grid()
            # Announcement of the final statement
            print("Congrats, you won!\n")
            break


while True:
    grid_clear()
    generate_pac_man()
    generate_ghost()
    generate_wall()
    play_game()
    c = input('Play Again [Y/N] ')
    if c not in 'yY':
        break

File: pac-man.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <cmath>
using namespace std;

const int N = 4;  //TWL
const int M = 8; //3RD
int counter=0;
char grid[N][M];

//This function prints the grid of Gomoku as the game progresses
void print_grid() {
    cout << "\n";
	cout << "--";
    for (int i = 0; i < M; cout << "---", i++);
    cout << "--\n";
    for (int i = 0; i < N; i++) {
        cout << "|  ";
        for (int j = 0; j < M; j++)
            cout << grid[i][j] << "  ";
        cout << "|\n";
        cout << "";
        for (int i = 0; i < M; cout << "", i++);
    }
    for (int i = 0; i < M; cout << "---", i++);
    cout << "----\n";
}

//This function checks if the game has a win state or not
bool check_win() {
    if (counter==N*M-(N*M*1/6 + N*M*1/8)-1) {return true;}
    else {return false;}
}
//This function checks if the game has a tie state or not for the given mark
bool check_lose(int i,int a , int b) {
if (i==3){  if(grid[a][b+1]=='G'){return true;}}

 else if (i==1){ if(grid[a-1][b]=='G'){return true;}}

 else if (i==2){  if(grid[a+1][b]=='G'){return true;}}

 else if (i==4){  if(grid[a][b-1]=='G'){return true;}}
  return false;
}

//This function checks if the game has a tie state or not for the given mark
bool check_wall(int i, int a , int b) {
if (i==3){  if(grid[a][b+1]=='|'){return true;}}

else if (i==1){ if(grid[a-1][b]=='|'){return true;}}

else if (i==2){  if(grid[a+1][b]=='|'){return true;}}

else if (i==4){  if(grid[a][b-1]=='|'){return true;}}
  return false;
}


//This function checks if given position is valid or not
bool check_available_move(int i,int &a,int &b) {
 if (i==3){  if(b+1>M-1){return false;}  } //right
 else if (i==1){  if(a-1<0){return false;} }   // up
 else if (i==2){  if(a+1>N-1){return false;}}   //down
 else if (i==4){  if(b-1<0){return false;}}     //left
return true;
 }


//This function checks if given position is valid or not
bool check_valid_direction(int i) {
    if (i==1 || i==2 || i==3 || i==4 ){
        return true;
        }
    else{return false;}
}

//This function generates pac man
void generate_pac_man(int &a , int &b) {
    srand( (unsigned)time( NULL ) );
    a = rand() % N;
	b = rand() % M;
    grid[a][b] = 'P';
}

//This function generates ghosts
void generate_ghost() {
    srand( (unsigned)time( NULL ) );
    int f=N*M*1/6;
    while ( f>0 )
    {
    int a = rand() % N;
	int b = rand() % M;
    while (grid[a][b] == 'P') {
        a = rand() % N;
        b = rand() % M;
	}
    grid[a][b] = 'G';
    f--;
    }
}

//This function generates a wall
void generate_wall() {
    srand( (unsigned)time( NULL ) );
    int f=N*M*1/8;
    while ( f>0 )
    {
    int a = rand() % N;
	int b = rand() % M;
    while (grid[a][b] == 'G' || grid[a][b] == 'P') {
        a = rand() % N;
        b = rand() % M;
	}
    grid[a][b] = '|';
    f--;
    }
}

//Move with the input direction
void move(int i,int &a,int &b) {
    grid[a][b]='.';
    if (i==3){  if(grid[a][b+1]=='+'){counter+=1; grid[a][b+1]='P';   }
                else if(grid[a][b+1]=='.'){ grid[a][b+1]='P';  } b+=1;  } //right

 else if (i==1){ if(grid[a-1][b]=='+'){counter+=1; grid[a-1][b]='P';}
                 else if(grid[a-1][b]=='.'){ grid[a-1][b]='P'; }  a-=1;  }   // up

 else if (i==2){  if(grid[a+1][b]=='+'){counter+=1; grid[a+1][b]='P';}
                    else if(grid[a+1][b]=='.'){ grid[a+1][b]='P'; } a+=1;}  //down

 else if (i==4){  if(grid[a][b-1]=='+'){counter+=1; grid[a][b-1]='P';}
                  else if(grid[a][b-1]=='.'){ grid[a][b-1]='P'; }b-=1;}  }   //left

//This function clears the game structures
void grid_clear() {
     for (int i =0 ; i<N; i++){
            for (int j =0 ; j<M; j++){
                grid[i][j]='+';
            }
        }
}
//This function reads a valid direction
void read_input(int &i,int &a, int &b) {
    cout << "Enter the direction: ";
	cin >> i;
    while (!check_valid_direction(i) || !check_available_move(i,a,b)) {
        cout << "Enter a valid direction: ";
		cin >> i;
	}
}
//MAIN FUNCTION
void play_game(int &a,int &b) {
    cout << "Pac-Man Game!\n";
    cout << "Welcome...\n";
    cout << "============================\n";
    while (true) {

        //Prints the grid
        print_grid();
        //Read an input from the player
		int i;
		read_input(i,a,b);

        //Check if the state of the grid has a tie state
        if (check_lose(i,a,b)) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "GAME OVER!\n";
            break;
        }
        if(check_wall(i,a,b)){
            cout<<"WHATCH OUT you hit a wall !! please " ;
            read_input(i,a,b);
        }
        //Move with the input direction
        move(i,a,b);
        //Check if the state of the grid has a win state
        if (check_win()) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Congrats, you won!\n";
            break;
        }

    }
}
int main() {
    while (true) {
        int a,b;
        grid_clear();
        //
        generate_pac_man(a,b);
        //
        generate_ghost();
        //
        generate_wall();
    	play_game(a,b);
    	char c;
    	cout << "Play Again [Y/N] ";
    	cin >> c;
    	if (c != 'y' && c != 'Y')
    		break;
    }
}
